File: hrms-frontend/lib/types/employee.ts
// Employee Management Types

export type Gender =
  | 'MALE'
  | 'FEMALE'
  | 'OTHER'
  | 'PREFER_NOT_TO_SAY';

export type EmploymentType =
  | 'FULL_TIME'
  | 'PART_TIME'
  | 'CONTRACT'
  | 'INTERN'
  | 'CONSULTANT';

export type EmployeeStatus =
  | 'ACTIVE'
  | 'ON_LEAVE'
  | 'ON_NOTICE'
  | 'TERMINATED'
  | 'RESIGNED';

export type EmployeeLevel =
  | 'ENTRY'
  | 'MID'
  | 'SENIOR'
  | 'LEAD'
  | 'MANAGER'
  | 'SENIOR_MANAGER'
  | 'DIRECTOR'
  | 'VP'
  | 'SVP'
  | 'CXO';

export type JobRole =
  // Engineering
  | 'SOFTWARE_ENGINEER'
  | 'FRONTEND_DEVELOPER'
  | 'BACKEND_DEVELOPER'
  | 'FULLSTACK_DEVELOPER'
  | 'DEVOPS_ENGINEER'
  | 'QA_ENGINEER'
  | 'DATA_ENGINEER'
  | 'MOBILE_DEVELOPER'
  | 'SYSTEM_ARCHITECT'
  | 'TECH_LEAD'
  | 'ENGINEERING_MANAGER'
  // Product
  | 'PRODUCT_MANAGER'
  | 'PRODUCT_OWNER'
  | 'PRODUCT_ANALYST'
  // Design
  | 'UI_DESIGNER'
  | 'UX_DESIGNER'
  | 'GRAPHIC_DESIGNER'
  | 'PRODUCT_DESIGNER'
  // Data & Analytics
  | 'DATA_ANALYST'
  | 'DATA_SCIENTIST'
  | 'BUSINESS_ANALYST'
  // Marketing
  | 'MARKETING_MANAGER'
  | 'CONTENT_WRITER'
  | 'SEO_SPECIALIST'
  | 'SOCIAL_MEDIA_MANAGER'
  | 'DIGITAL_MARKETER'
  // Sales
  | 'SALES_REPRESENTATIVE'
  | 'SALES_MANAGER'
  | 'ACCOUNT_MANAGER'
  | 'BUSINESS_DEVELOPMENT'
  // Operations
  | 'OPERATIONS_MANAGER'
  | 'PROJECT_MANAGER'
  | 'SCRUM_MASTER'
  | 'PROGRAM_MANAGER'
  // HR
  | 'HR_MANAGER'
  | 'HR_GENERALIST'
  | 'RECRUITER'
  | 'TALENT_ACQUISITION'
  // Finance
  | 'ACCOUNTANT'
  | 'FINANCIAL_ANALYST'
  | 'FINANCE_MANAGER'
  // Admin & Support
  | 'ADMIN_ASSISTANT'
  | 'OFFICE_MANAGER'
  | 'CUSTOMER_SUPPORT'
  | 'TECH_SUPPORT'
  // Legal
  | 'LEGAL_COUNSEL'
  | 'COMPLIANCE_OFFICER'
  // Other
  | 'CONSULTANT'
  | 'INTERN'
  | 'OTHER';

export interface Employee {
  id: string;
  userId?: string;
  employeeCode: string;
  firstName: string;
  middleName?: string;
  lastName?: string;
  fullName: string;
  workEmail: string;
  personalEmail?: string;
  phoneNumber?: string;
  emergencyContactNumber?: string;
  dateOfBirth?: string; // YYYY-MM-DD
  gender?: Gender;
  address?: string;
  city?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  joiningDate: string; // YYYY-MM-DD
  confirmationDate?: string; // YYYY-MM-DD
  exitDate?: string; // YYYY-MM-DD
  departmentId?: string;
  departmentName?: string;
  designation?: string;
  managerId?: string;
  managerName?: string;
  employmentType: EmploymentType;
  status: EmployeeStatus;
  bankAccountNumber?: string;
  bankName?: string;
  bankIfscCode?: string;
  taxId?: string;
  createdAt: string;
  updatedAt: string;
  level?: EmployeeLevel;
  jobRole?: JobRole;
  subordinates?: Employee[];
}

export interface CreateEmployeeRequest {
  employeeCode: string;
  firstName: string;
  middleName?: string;
  lastName?: string;
  workEmail: string;
  password?: string;
  personalEmail?: string;
  phoneNumber?: string;
  emergencyContactNumber?: string;
  dateOfBirth?: string; // YYYY-MM-DD
  gender?: Gender;
  address?: string;
  city?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  joiningDate: string; // YYYY-MM-DD
  confirmationDate?: string; // YYYY-MM-DD
  departmentId?: string;
  designation?: string;
  managerId?: string;
  employmentType: EmploymentType;
  status: EmployeeStatus;
  bankAccountNumber?: string;
  bankName?: string;
  bankIfscCode?: string;
  taxId?: string;
  level?: EmployeeLevel;
  jobRole?: JobRole;
}

export interface UpdateEmployeeRequest {
  firstName?: string;
  middleName?: string;
  lastName?: string;
  personalEmail?: string;
  phoneNumber?: string;
  emergencyContactNumber?: string;
  dateOfBirth?: string; // YYYY-MM-DD
  gender?: Gender;
  address?: string;
  city?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  confirmationDate?: string; // YYYY-MM-DD
  departmentId?: string;
  designation?: string;
  managerId?: string;
  employmentType?: EmploymentType;
  status?: EmployeeStatus;
  bankAccountNumber?: string;
  bankName?: string;
  bankIfscCode?: string;
  taxId?: string;
  level?: EmployeeLevel;
  jobRole?: JobRole;
}

export interface Page<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

// Department Types

export type DepartmentType =
  | 'ENGINEERING'
  | 'PRODUCT'
  | 'DESIGN'
  | 'MARKETING'
  | 'SALES'
  | 'OPERATIONS'
  | 'FINANCE'
  | 'HR'
  | 'LEGAL'
  | 'ADMIN'
  | 'SUPPORT'
  | 'OTHER';

export interface Department {
  id: string;
  code: string;
  name: string;
  description?: string;
  parentDepartmentId?: string;
  parentDepartmentName?: string;
  managerId?: string;
  managerName?: string;
  isActive: boolean;
  location?: string;
  costCenter?: string;
  type?: DepartmentType;
  employeeCount?: number;
  subDepartmentCount?: number;
  createdAt: string;
  updatedAt: string;
  subDepartments?: Department[]; // For hierarchical structure
}

export interface DepartmentRequest {
  code: string;
  name: string;
  description?: string;
  parentDepartmentId?: string;
  managerId?: string;
  isActive?: boolean;
  location?: string;
  costCenter?: string;
  type?: DepartmentType;
}

// Employee Import Types

export type ImportErrorType =
  | 'REQUIRED_FIELD_MISSING'
  | 'INVALID_FORMAT'
  | 'INVALID_VALUE'
  | 'DUPLICATE_IN_FILE'
  | 'DUPLICATE_IN_DATABASE'
  | 'REFERENCE_NOT_FOUND'
  | 'BUSINESS_RULE_VIOLATION';

export interface ImportValidationError {
  rowNumber: number;
  field: string;
  value?: string;
  message: string;
  errorType: ImportErrorType;
}

export interface EmployeeImportRowPreview {
  rowNumber: number;
  employeeCode: string;
  fullName: string;
  workEmail: string;
  designation: string;
  departmentName?: string;
  joiningDate: string;
  employmentType: string;
  isValid: boolean;
  rowErrors: string[];
}

export interface EmployeeImportPreview {
  totalRows: number;
  validRows: number;
  invalidRows: number;
  hasErrors: boolean;
  rows: EmployeeImportRowPreview[];
  errors: ImportValidationError[];
}

export type ImportStatus =
  | 'COMPLETED'
  | 'PARTIAL_SUCCESS'
  | 'FAILED';

export interface ImportedEmployee {
  rowNumber: number;
  employeeId: string;
  employeeCode: string;
  fullName: string;
  workEmail: string;
}

export interface FailedImport {
  rowNumber: number;
  employeeCode?: string;
  reason: string;
  errors?: ImportValidationError[];
}

export interface EmployeeImportResult {
  importId: string;
  importedAt: string;
  importedBy: string;
  totalProcessed: number;
  successCount: number;
  failedCount: number;
  skippedCount?: number;
  status: ImportStatus;
  importedEmployees?: ImportedEmployee[];
  failedImports?: FailedImport[];
}

File: backend/src/main/java/com/hrms/domain/employee/Employee.java
package com.hrms.domain.employee;

import com.hrms.common.entity.TenantAware;
import com.hrms.domain.user.User;
import jakarta.persistence.*;
import org.hibernate.annotations.Where;
import lombok.*;
import lombok.experimental.SuperBuilder;

import java.time.LocalDate;
import java.util.UUID;

@Where(clause = "is_deleted = false")
@Entity
@Table(name = "employees", indexes = {
        @Index(name = "idx_employee_code_tenant", columnList = "employeeCode,tenantId", unique = true),
        @Index(name = "idx_employee_tenant", columnList = "tenantId"),
        @Index(name = "idx_employee_department", columnList = "departmentId"),
        @Index(name = "idx_employee_manager", columnList = "managerId"),
        @Index(name = "idx_employee_dotted_mgr1", columnList = "dottedLineManager1Id"),
        @Index(name = "idx_employee_dotted_mgr2", columnList = "dottedLineManager2Id"),
        @Index(name = "idx_employee_status", columnList = "status")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
public class Employee extends TenantAware {

    @Column(nullable = false, length = 50)
    private String employeeCode;

    @OneToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 100)
    private String firstName;

    @Column(length = 100)
    private String middleName;

    @Column(length = 100)
    private String lastName;

    @Column(length = 200)
    private String personalEmail;

    @Column(length = 20)
    private String phoneNumber;

    @Column(length = 20)
    private String emergencyContactNumber;

    @Column
    private LocalDate dateOfBirth;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    private Gender gender;

    @Column(columnDefinition = "TEXT")
    private String address;

    @Column(length = 100)
    private String city;

    @Column(length = 100)
    private String state;

    @Column(length = 20)
    private String postalCode;

    @Column(length = 100)
    private String country;

    @Column(name = "avatar_url", length = 500)
    private String avatarUrl;

    @Column(nullable = false)
    private LocalDate joiningDate;

    @Column
    private LocalDate confirmationDate;

    @Column
    private LocalDate exitDate;

    @Column
    private UUID departmentId;

    @Column
    private UUID officeLocationId;

    @Column
    private UUID teamId;

    @Column(length = 100)
    private String designation;

    @Enumerated(EnumType.STRING)
    @Column(length = 50)
    private EmployeeLevel level;

    @Enumerated(EnumType.STRING)
    @Column(length = 50)
    private JobRole jobRole;

    @Column
    private UUID managerId;

    /** Optional first dotted-line manager (matrix reporting, informational only). */
    @Column
    private UUID dottedLineManager1Id;

    /** Optional second dotted-line manager (matrix reporting, informational only). */
    @Column
    private UUID dottedLineManager2Id;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private EmploymentType employmentType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private EmployeeStatus status;

    @Column(length = 100)
    private String bankAccountNumber;

    @Column(length = 50)
    private String bankName;

    @Column(length = 50)
    private String bankIfscCode;

    @Column(length = 50)
    private String taxId;

    public enum Gender {
        MALE,
        FEMALE,
        OTHER,
        PREFER_NOT_TO_SAY
    }

    public enum EmploymentType {
        FULL_TIME,
        PART_TIME,
        CONTRACT,
        INTERN,
        CONSULTANT
    }

    public enum EmployeeStatus {
        ACTIVE,
        INACTIVE,
        ON_LEAVE,
        ON_NOTICE,
        TERMINATED,
        RESIGNED
    }

    public enum EmployeeLevel {
        ENTRY, // Junior/Associate
        MID, // Mid-level
        SENIOR, // Senior
        LEAD, // Lead/Principal
        MANAGER, // Manager
        SENIOR_MANAGER, // Senior Manager
        DIRECTOR, // Director
        VP, // Vice President
        SVP, // Senior Vice President
        CXO // C-level Executive
    }

    public enum JobRole {
        // Executive
        EXECUTIVE,
        CEO,
        CTO,
        CFO,
        COO,

        // Engineering
        ENGINEERING,
        SOFTWARE_ENGINEER,
        FRONTEND_DEVELOPER,
        BACKEND_DEVELOPER,
        FULLSTACK_DEVELOPER,
        DEVOPS_ENGINEER,
        QA_ENGINEER,
        DATA_ENGINEER,
        MOBILE_DEVELOPER,
        SYSTEM_ARCHITECT,
        TECH_LEAD,
        ENGINEERING_MANAGER,

        // Product
        PRODUCT_MANAGER,
        PRODUCT_OWNER,
        PRODUCT_ANALYST,

        // Design
        UI_DESIGNER,
        UX_DESIGNER,
        GRAPHIC_DESIGNER,
        PRODUCT_DESIGNER,

        // Data & Analytics
        DATA_ANALYST,
        DATA_SCIENTIST,
        BUSINESS_ANALYST,

        // Marketing
        MARKETING_MANAGER,
        CONTENT_WRITER,
        SEO_SPECIALIST,
        SOCIAL_MEDIA_MANAGER,
        DIGITAL_MARKETER,

        // Sales
        SALES_REPRESENTATIVE,
        SALES_MANAGER,
        ACCOUNT_MANAGER,
        BUSINESS_DEVELOPMENT,

        // Operations
        OPERATIONS_MANAGER,
        PROJECT_MANAGER,
        SCRUM_MASTER,
        PROGRAM_MANAGER,

        // HR
        HR_MANAGER,
        HR_GENERALIST,
        RECRUITER,
        TALENT_ACQUISITION,
        HUMAN_RESOURCES, // Legacy value from seed data

        // Finance
        ACCOUNTANT,
        FINANCIAL_ANALYST,
        FINANCE_MANAGER,
        FINANCE, // Legacy value from seed data

        // Admin & Support
        ADMIN_ASSISTANT,
        OFFICE_MANAGER,
        CUSTOMER_SUPPORT,
        TECH_SUPPORT,

        // Legal
        LEGAL_COUNSEL,
        COMPLIANCE_OFFICER,

        // Other
        CONSULTANT,
        INTERN,
        OTHER
    }

    public String getFullName() {
        StringBuilder name = new StringBuilder();
        if (firstName != null) {
            name.append(firstName);
        }
        if (middleName != null && !middleName.isEmpty()) {
            if (name.length() > 0) name.append(" ");
            name.append(middleName);
        }
        if (lastName != null && !lastName.isEmpty()) {
            if (name.length() > 0) name.append(" ");
            name.append(lastName);
        }
        return name.length() > 0 ? name.toString() : "";
    }

    public void terminate() {
        this.status = EmployeeStatus.TERMINATED;
        this.exitDate = LocalDate.now();
    }

    public void resign() {
        this.status = EmployeeStatus.RESIGNED;
        this.exitDate = LocalDate.now();
    }
}
